File: benchmarks/wrappers/gcp/python/storage.py
import io
import json
import os
import uuid

from google.cloud import storage as gcp_storage


class storage:
    instance = None
    client = None

    def __init__(self):
        self.client = gcp_storage.Client()

    @staticmethod
    def unique_name(name):
        name, extension = os.path.splitext(name)
        return '{name}.{random}{extension}'.format(
                    name=name,
                    extension=extension,
                    random=str(uuid.uuid4()).split('-')[0]
                )

    def upload(self, bucket, file, filepath, overwrite=False):
        key_name = storage.unique_name(file)
        if (overwrite):
            key_name = file
        bucket_instance = self.client.bucket(bucket)
        blob = bucket_instance.blob(key_name)
        blob.upload_from_filename(filepath)
        return key_name

    def download(self, bucket, file, filepath):
        bucket_instance = self.client.bucket(bucket)
        blob = bucket_instance.blob(file)
        blob.download_to_filename(filepath)

    def download_directory(self, bucket, prefix, path):
        objects = self.client.bucket(bucket).list_blobs(prefix=prefix)
        for obj in objects:
            file_name = obj.name
            path_to_file = os.path.dirname(file_name)
            os.makedirs(os.path.join(path, path_to_file), exist_ok=True)
            self.download(bucket, file_name, os.path.join(path, file_name))

    def upload_stream(self, bucket, file, data, overwrite=False):
        key_name = storage.unique_name(file)
        if (overwrite):
            key_name = file
        bucket_instance = self.client.bucket(bucket)
        blob = bucket_instance.blob(key_name)
        blob.upload_from_file(data)
        return key_name

    def download_stream(self, bucket, file):
        data = io.BytesIO()
        bucket_instance = self.client.bucket(bucket)
        blob = bucket_instance.blob(file)
        blob.download_to_file(data)
        return data.getbuffer()

    def get_object(self, bucket, key):
        bucket_instance = self.client.bucket(bucket)
        blob = bucket_instance.blob(key)
        contents = blob.download_as_bytes()
        return contents

    def get_instance():
        if storage.instance is None:
            storage.instance = storage()
        return storage.instance

    def list_objects(self, bucket, prefix=None):
        res = self.client.list_blobs(bucket, prefix=prefix)

        objs = []
        for obj in res:
            objs.append(obj.name)

        return objs

    def delete_object(self, bucket, key):
        bucket = self.client.bucket(bucket)
        blob = bucket.blob(key)
        blob.delete()

File: benchmarks/wrappers/gcp/nodejs/storage.js
// Copyright 2020-2025 ETH Zurich and the SeBS authors. All rights reserved.
const { Storage } = require('@google-cloud/storage'),
        fs = require('fs'),
        path = require('path'),
        uuid = require('uuid'),
        util = require('util'),
        stream = require('stream');

class gcp_storage {

  constructor() {
    this.storage = new Storage();
  }

  unique_name(file) {
    let name = path.parse(file);
    let uuid_name = uuid.v4().split('-')[0];
    return path.join(name.dir, util.format('%s.%s%s', name.name, uuid_name, name.ext));
  }

  upload(container, file, filepath) {
    let bucket = this.storage.bucket(container);
    let uniqueName = this.unique_name(file);
    let options = {destination: uniqueName, resumable: false};
    return [uniqueName, bucket.upload(filepath, options)];
  };

  download(container, file, filepath) {
    let bucket = this.storage.bucket(container);
    var file = bucket.file(file);
    return file.download({destination: filepath});
  };

  async downloadDirectory(container, prefix, downloadPath) {
    let bucket = this.storage.bucket(container);
    const [files] = await bucket.getFiles({ prefix: prefix });

    const downloadPromises = files.map(file => {
      const fileName = file.name;
      const pathToFile = path.dirname(fileName);
      fs.mkdirSync(path.join(downloadPath, pathToFile), { recursive: true });
      return this.download(container, fileName, path.join(downloadPath, fileName));
    });

    await Promise.all(downloadPromises);
  };

  uploadStream(container, file) {
    let bucket = this.storage.bucket(container);
    let uniqueName = this.unique_name(file);
    var file = bucket.file(uniqueName);
    let upload = file.createWriteStream();
    var write_stream = new stream.PassThrough();

    write_stream.pipe(upload);

    const promise = new Promise((resolve, reject) => {
      upload.on('error', err => {
        upload.end();
        reject(err);
      });

      upload.on('finish', () => {
        upload.end();
        resolve(file.name);
      });
    });
    return [write_stream, promise, uniqueName];
  };

  downloadStream(container, file) {
    let bucket = this.storage.bucket(container);
    var file = bucket.file(file);
    let downloaded = file.createReadStream();
    return Promise.resolve(downloaded);
  };
};
exports.storage = gcp_storage;
